linked_list/main.py

```diff
@@ -62,34 +68,116 @@ class LinkedList(AbstractLinkedList):
     """
 
     def __init__(self, elements=None):
-        pass
+        self.start = None
+        self.end = None
+        
+        if elements:
+            
+            self.end = Node(elements[-1])
+            node = Node(elements[0], None)
+            last_node = self.end
+            for element in reversed(elements[:-1]):
+                node = Node(element, last_node)
+                last_node = node
+            self.start = node
+    
 
     def __str__(self):
-        pass
-
+        lst = [str(el) for el in self]
+        
+        return "[" + ", ".join(lst) + "]"
+        
     def __len__(self):
-        pass
+        return self.count()
 
     def __iter__(self):
-        pass
+        return (node.elem for node in self.iterate_nodes())
+        
 
     def __getitem__(self, index):
-        pass
+        if index < 0:
+            index = len(self) + index
+            
+        for count, elem in enumerate(self):
+            if count == index:
+                return elem
+
+            
 
     def __add__(self, other):
-        pass
+        
+        new_list = [el for el in self]
+        new_list += [el for el in other]
+
+        return LinkedList(new_list)
 
     def __iadd__(self, other):
-        pass
+        
+        for el in other:
+            self.append(el)
+            
+        return self
 
     def __eq__(self, other):
-        pass
 
+        for el1, el2 in zip(self, other):
+            if el1 != el2:
+                return False
+                
+        return len(self) == len(other)
+                
+    def iterate_nodes(self):
+        if self.start:
+            node = self.start
+            while node:
+                yield node
+                node = node.next
+                
     def append(self, elem):
-        pass
+        node = Node(elem)
+        if self.end:
+            self.end.next = node
+        else:
+            self.start = node
+            
+        self.end = node
+
+
 
     def count(self):
-        pass
+        count = 0
+        for _ in self:
+            count += 1
+        return count
+
 
     def pop(self, index=None):
-        pass
+    
+        if index is None:
+            index = len(self) - 1
+    
+        if self.end is None or index >= len(self):
+            raise IndexError()
+            
+        output = self[index]
+        
+        if index == 0:
+           self.start = self.start.next
+        else:
+            node_before_pop = self.get_node(index - 1)
+            node_to_be_popped = self.get_node(index)
+            node_after_pop =  node_to_be_popped.next
```

Comment:
You really only need to get 1 node for the pop. The one proceeding (index - 1), then using the `next()` method you can find the following elements. Remember it can be chained (`node.next().next()').